lib/utils/extract.js

```diff
@@ -0,0 +1,223 @@
+ */
+function getFullUrl(url) {
+  if (url.startsWith(prefix)) return url;
+  return `${prefix}${url}`;
+}
+
+/**
+ * Returns the extension on the URL or '' if there is none.
+ * @param {string} url The URL.
+ * @return {string} The URL's extension or the empty string if the URL has no
+ *     extension.
+ */
+function getExtension(url) {
+  const splits = url.split('.');
+  let extension = splits[splits.length - 1].trim();
+  if (!/^[a-zA-Z0-9]+$/.test(extension) || extension.length > 10) {
+    // Clear the extension if it is not alphanumeric or is long enough to
+    // signify it may just be a hash value or something.
+    extension = '';
+  }
+  return extension;
+}
+
+/**
+ * Returns the relative path for the specified URL, adding index.html if the URL
+ * ends in `/`. This makes the URL function well in a static site.
+ * @param {string} url The URL for which to retrieve the relative path.
+ * @return {string} A URL that starts with / that is relative to the server
+ *     root. The URL will add index.html if it ends with `/`.
+ */
+function relativePath(url) {
+  if (url.startsWith(prefix)) return relativePath(url.substr(prefix.length));
+  if (url.endsWith('/')) url += 'index.html';
+  if (getExtension(url) == '') url += '/index.html';
+  if (url.startsWith('/')) return url;
+  throw new Error('Bad url');
+}
+
+/**
+ * Returns the Sapper API route for the specified URL path.
+ * @param {string} url The absolute or relative URL.
+ * @return {string} The URL with /api/ in front.
+ */
+function apiPath(url) {
+  if (url.startsWith(prefix)) {
+    return `${prefix}/api${url.substr(prefix.length)}`;
+  }
+  return `/api${url}`;
+}
+
+/**
+ * Returns whether the specified URL is on the server or an external link.
+ * @param {string} url The URL.
+ * @return {boolean} True if the URL is on the server.
+ */
+function filter(url) {
+  return url.startsWith('/') || url.startsWith(getFullUrl('/'));
+}
+
+/**
+ * Retrieves chunk files that are normally cached for offline use in the service
+ * worker.
+ * @return {!Array<string>}
+ */
+function getChunkFiles() {
+  const clientInfo =
+      fs.readJsonSync(path.join(sapperDest, 'stats.client.json'));
+  const chunkFiles = clientInfo.assets.map(chunk => `/client/${chunk.name}`);
+  return chunkFiles;
+}
+
+/**
+ * Exports the Sapper app as a static website by starting at the root and
+ * crawling pages that are linked, their /api/ pages, and webpack routes, as
+ * well as copying assets.
+ * @param {?Array<string>=} includeUrls If non-null, a set of additional URLs to
+ *     scrape in the extraction. This should only be set if there are routes
+ *     that cannot be reached from the root.
+ * @param {?Array<string>=} excludeUrls If non-null, a set of URLs to avoid
+ *     scraping in the extraction.
+ * @param {number=} extractionDir The directory in which to place the extracted
+ *     output.
+ */
+module.exports = async function(includeUrls = null, excludeUrls = null,
```

Comment:
Super cool stuff! There is an [issue to move away from `async/await`](https://github.com/sveltejs/sapper/issues/62). Maybe it's best to preemptively not use it?

Reply:
Thanks! Hmm, there are a lot of `await` calls in Sapper already. I'm happy to change out my `await` and `async` calls, but do you think the onus of not being able to run on platforms like AWS Lambda because of the Node version is on the platforms to fix?

<img width="337" alt="await_in_sapper" src="https://user-images.githubusercontent.com/306095/34631263-69988a1e-f224-11e7-9feb-317ca27bc049.png">

Reply:
@Rich-Harris has more insight on that than me, but I can definitely see the appeal of supporting lower versions of Node. 
I just have all the `async/await` in `test.js` left to replace in a PR, so no worries!

Reply:
I've removed `async/await` in my changes to keep you from having to do the work :)